routes/api.js

```diff
@@ -84,24 +88,103 @@ var substituteCodeCell = function(payload) {
             transformedData = '';
         }
     }
-    return transformedData;
+    return Promise.resolve(transformedData);
 };
 
 // reusable objects required by WebSocketFrame
-var maskBytes = new Buffer(4);
-var frameHeader = new Buffer(10);
-var bufferList = new BufferList();
-var wsconfig = {
-    maxReceivedFrameSize: 0x100000  // 1MiB max frame size.
-};
+// var maskBytes = new Buffer(4);
+// var frameHeader = new Buffer(10);
+// var bufferList = new BufferList();
+// var wsconfig = {
+//     maxReceivedFrameSize: 0x100000  // 1MiB max frame size.
+// };
+
+// var wsconfig = {
+//     maxReceivedFrameSize: Math.pow(2, 64)
+// };
+
+function sendSocketData(conn, data) {
+    if (data.type === 'utf8') {
+        conn.sendUTF(data.utf8Data);
+    } else {
+        conn.sendBytes(data.binaryData);
+    }
+}
 
 function setupWSProxy(_server) {
     debug('setting up WebSocket proxy');
     server = _server;
 
+
+    var wsserver = new WebSocketServer({
+        httpServer: _server,
+        autoAcceptConnections: false,
+        maxReceivedFrameSize: Number.MAX_SAFE_INTEGER
+    });
+
+    wsserver.on('request', function(request) {
+        debug('ws connection request');
+
+        // XXX verify request.origin
+        var servConn = request.accept(null, request.origin);
+
+        var pendingServMsgs = [];
+        var bufferServMsgs = function(data) {
+            pendingServMsgs.push(data);
+        };
+        servConn.on('message', bufferServMsgs);
+        servConn.on('close', function(reasonCode, desc) {
+            debug('closing WS server connection:', reasonCode, desc);
```

Comment:
Probably need to move the kernel deletion timer and callback here. Doesn't seem to fire anymore on server close.